src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SamAuthProvider.scala

```diff
@@ -0,0 +1,218 @@
+    Future{ resourcesApiAsPet(userEmail, googleProject).resourceAction(billingProjectResourceTypeName, googleProject, getProjectActionString(action)) }
+  }
+
+  /**
+    * Leo calls this method to verify if the user has permission to perform the given action on a specific notebook cluster.
+    * It may call this method passing in a cluster that doesn't exist. Return Future.successful(false) if so.
+    *
+    * @param userEmail      The user in question
+    * @param action        The cluster-level action (above) the user is requesting
+    * @param googleProject The Google project the cluster was created in
+    * @param clusterName   The user-provided name of the Dataproc cluster
+    * @return If the userEmail has permission on this individual notebook cluster to perform this action
+    */
+  def hasNotebookClusterPermission(userEmail: WorkbenchEmail, action: NotebookClusterActions.NotebookClusterAction, googleProject: String, clusterName: String)(implicit executionContext: ExecutionContext): Future[Boolean] = {
+    // get the id for the cluster resource
+    val clusterResourceId = getClusterResourceId(googleProject, clusterName)
+
+    // if action is connect, check only cluster resource. If action is anything else, either cluster or project must be true
+    Future {
+      val notebookAction = resourcesApiAsPet(userEmail, googleProject).resourceAction(notebookClusterResourceTypeName, clusterResourceId, getNotebookClusterActionString(action))
+      if (action == ConnectToCluster) {
+        notebookAction
+      } else {
+        notebookAction || resourcesApiAsPet(userEmail, googleProject).resourceAction(billingProjectResourceTypeName, googleProject, getNotebookClusterActionString(action))
+      }
+    }
+  }
+
+  //Notifications that Leo has created/destroyed clusters. Allows the auth provider to register things.
+
+  /**
+    * Leo calls this method to notify the auth provider that a new notebook cluster has been created.
+    * The returned future should complete once the provider has finished doing any associated work.
+    * Returning a failed Future will prevent the cluster from being created, and will call notifyClusterDeleted for the same cluster.
+    * Leo will wait, so be timely!
+    *
+    * @param userEmail      The email address of the user in question
+    * @param googleProject The Google project the cluster was created in
+    * @param clusterName   The user-provided name of the Dataproc cluster
+    * @return A Future that will complete when the auth provider has finished doing its business.
+    */
+  def notifyClusterCreated(userEmail: WorkbenchEmail, googleProject: String, clusterName: String)(implicit executionContext: ExecutionContext): Future[Unit] = {
+    val clusterResourceId = getClusterResourceId(googleProject, clusterName)
+    // Add the cluster resource with the user as owner
+    Future { resourcesApiAsPet(userEmail, googleProject).createResource(notebookClusterResourceTypeName, clusterResourceId) }
```

Comment:
I think we need to doublecheck the logic between LeoService, SamAuthProvider, and what we put into Sam. There's a `DeleteClusters` project action that isn't used by `LeoService`; plus `LeoService` has a tendency to check only `GetClusterStatus` to see if a user can know about a cluster when it should also be testing `ListClusters` too.

Let's take some time to go through this @ansingh7115 @rtitle